incubator/virtualcluster/doc/customresource-syncer.md

```diff
@@ -0,0 +1,185 @@
+       "sigs.k8s.io/controller-runtime/pkg/cache")
+
+ 
+ superFooClient, err := client.New(config.RestConfig, client.Options{})
+ 
+superFoocache, err = cache.New(config.RestConfig, cache.Options{})
+superFooInformer, err := c.superFoocache.GetInformer(context.Background(), &alpha1.Foo{})
+  
+```
+
+CR cache needs to be bootstrapped, We can put the bootstrap logic in DWS or UWS initialization function as:
+
+```
+func (c *controller) StartDWS(stopCh <-chan struct{}) error {
+go superFoocache.Start(stopCh)
+}
+```
+
+
+For Virtual Cluster CR Client, we can either get informer from CR specific Mccontroller, or construct client using tenant virtual cluster Restful config.
+
+```
+ clusterInformer, err := c.multiClusterFooController.GetCluster(cluster).GetInformer(&alpha1.foo{})
+ 
+ vcRestConfig := c.MultiClusterController.GetCluster(vclusterName).GetRestConfig()
+ virtualFooClient, err := client.New(vcRestConfig, client.Options{})
+```
+
+### CR Scheme Building
+
+Custom Resource Scheme installation happens in Init function of each CR Syncer package.  Controller runtime scheme builder is used here.
+
+```
+import ( 
+    "k8s.io/apimachinery/pkg/runtime"
+    "k8s.io/apimachinery/pkg/runtime/schema"
+    )
+var SchemeGroupVersion = schema.GroupVersion{
+        Group:   "foo.example.com",
+        Version: "v1alpha1",
+}
+var (
+        SchemeBuilder      runtime.SchemeBuilder
+        localSchemeBuilder = &SchemeBuilder
+)
+func init() {
+        localSchemeBuilder.Register(addKnownTypes)
+        localSchemeBuilder.AddToScheme(scheme.Scheme)
+}
+```
+
+
+Foo CRD should be added into virtualcluster/pkg/syncer/util/scheme  to enable CR object conversion in Mccontroller:
+
+```
+ import (utilscheme "virtualcluster/pkg/syncer/util/scheme")
+ 
+ utilscheme.Scheme.AddKnownTypePair(&alpha1.Foo{}, &alpha1.FooList{})
+ 
+```
+
+## CR Syncer Testing 
+
+CR syncer relies on self constructed CR clients to perform synchronization between virtual and super clusters . Current testing framework in multi-tenancy/incubator/virtualcluster/pkg/syncer/util/test uses /k8s.io/client-go/ fake client to initialize resource controllers:
+
+```
+func NewFooController(config *config.SyncerConfiguration,
+        client clientset.Interface,
+        informer informers.SharedInformerFactory,
+        vcClient vcclient.Interface,
+        vcInformer vcinformers.VirtualClusterInformer,
+        options manager.ResourceSyncerOptions)
+```
+
+/k8s.io/client-go client cannot be extended to embed CR client. Therefore a different different NewFooController will be built to pass in CR fake client/informer instances to CR Syncer. 
+
+When multi-tenancy code base is migrated  from /[k8s.io/client-go/](http://k8s.io/client-go/) client  to [sigs.k8s.io/controller-runtime/](http://sigs.k8s.io/controller-runtime/) client, CR client can be included in client interface,  Multi-tenancy test framework can then be used for CR Syncer testing.
```

Comment:
I would remove this for now since it's not scoped work.

Reply:
agree